metronome/checkpointing/models/src/io/iohk/metronome/checkpointing/models/CheckpointCertificate.scala

```diff
@@ -31,14 +34,65 @@ case class CheckpointCertificate(
 )
 
 object CheckpointCertificate {
+
+  /** Create a `CheckpointCertificate` from a `Block` that last had a `CheckpointCandidate`
+    * and a list of `Block.Header`s leading up to the `QuorumCertifictate` that proves the
+    * BFT agreement over the contents.
+    */
   def construct(
       block: Block,
       headers: NonEmptyList[Block.Header],
       commitQC: QuorumCertificate[CheckpointingAgreement, Phase.Commit]
-  ): Option[CheckpointCertificate] =
+  ): Option[CheckpointCertificate] = {
+    assert(block.hash == headers.head.hash)
```

Comment:
So we intentionally do not return information that block header do not match (return Option rather Either)?

Reply:
You're asking why there's an `assert` rather than a "error" return value like `Left`, right?

This assertion is just to make sure we're calling this method with the right data. It would be a programming error if we passed something else, so this just puts into code what's in the comments, to catch simple ordering errors early during development. In practice we should never call it with invalid data, so there's no reason to be prepared for an error to be returned due to these reasons.

Reply:
Basically just sanity checking that [these comments](https://github.com/input-output-hk/metronome/blob/f2d70763dc36e37bd18c851e5d1707c255f094f3/metronome/checkpointing/models/src/io/iohk/metronome/checkpointing/models/CheckpointCertificate.scala#L25-L33) are taken into account.